[ADBDEV-8588] Add basic shrink/rebalance algorithm

bimboterminator1 opened the pull request:
This patch adds foundations of shrink/rebalance planner. Some extra planning
details and proper integration of planning stage into the ggrebalance state
machine are going to be considered in separate tickets.

The main feature of provided code is an abstract balancing algorithm, which
represents manual primary/mirror host assignment following greedy strategy.
In short, algorithm structure consists of several phases:

1) Primary assignment. Sort segments by relocation priority: firstly, must-move
segments - those lying at decomissioned hosts, encoded in initial_primary as
indexes >= n_target_hosts. Then move from overloaded to underloaded hosts.
Assign each segment to least-loaded host, preferring original placement when
possible.

2) Mirror assignment. Is built according to simple logic: prefer original
mirror hosts, use least-loaded mirror hosts.

3) Optional improvement. Using adaptive large neighborhood search, where we try
build near solutions by destroying and reassigning parts of the initial one.
Quite volatile, but in some cases can bring better solution. Proposed to use
in the ggrebalance utility. Reentrancy could be achieved by saving first plan
into the database.

Unit tests are moved from gppylib into gprebalance_modules in order to achieve
better tests granularity and possibility to import separate modules.

### How to test solver.py manually:

<details>

<summary>config generator</summary>


```python
   import random
from typing import List, Tuple

class ConfigGenerator:
    """Generate test configurations for segment placement"""

    @staticmethod
    def generate_balanced_grouped(n_segments: int, n_hosts: int) -> Tuple[List[int], List[int]]:
        """Generate balanced GROUPED configuration
        
        Balanced: Each host has equal number of primaries
        Grouped: All primaries on host_i have mirrors on host_(i+1)
        """
        assert n_segments % n_hosts == 0, "Must be evenly divisible"

        segs_per_host = n_segments // n_hosts

        # Assign primaries round-robin
        primaries = []
        for host_id in range(n_hosts):
            primaries.extend([host_id] * segs_per_host)

        # Assign mirrors (grouped: all primaries on host_i → mirrors on host_(i+1))
        mirrors = []
        for seg_id in range(n_segments):
            p_host = primaries[seg_id]
            m_host = (p_host + 1) % n_hosts
            mirrors.append(m_host)

        return primaries, mirrors

    @staticmethod
    def generate_balanced_spread(n_segments: int, n_hosts: int) -> Tuple[List[int], List[int]]:
        """Generate balanced SPREAD configuration
        
        Balanced: Each host has equal number of primaries
        Spread: Mirrors distributed evenly across all other hosts
        """
        assert n_segments % n_hosts == 0, "Must be evenly divisible"

        segs_per_host = n_segments // n_hosts

        # Assign primaries
        primaries = []
        for host_id in range(n_hosts):
            primaries.extend([host_id] * segs_per_host)

        # Assign mirrors (spread: distribute across different hosts)
        mirrors = []
        for host_id in range(n_hosts):
            # Available mirror hosts (not self)
            available = [h for h in range(n_hosts) if h != host_id]

            for i in range(segs_per_host):
                m_host = available[i % len(available)]
                mirrors.append(m_host)

        return primaries, mirrors

    @staticmethod
    def generate_unbalanced_grouped(n_segments: int, n_hosts: int, 
                                   skew: float = 0.3) -> Tuple[List[int], List[int]]:
        """Generate UNBALANCED GROUPED configuration
        
        Unbalanced: Skewed primary distribution (some hosts have more segments)
        Grouped: Mirrors follow grouped pattern (p_host + 1)
        
        Args:
            skew: Fraction of segments to transfer (0.3 = 30% imbalance)
        """
        base = n_segments // n_hosts

        # Create skewed distribution
        loads = [base] * n_hosts

        # Transfer segments from last hosts to first hosts
        transfer = int(base * skew)
        if transfer > 0 and n_hosts >= 2:
            loads[0] += transfer
            loads[-1] -= transfer

        # Assign primaries
        primaries = []
        for host_id in range(n_hosts):
            primaries.extend([host_id] * loads[host_id])

        # Shuffle to randomize segment order
        random.shuffle(primaries)

        # Assign mirrors (grouped pattern)
        mirrors = []
        for seg_id in range(n_segments):
            p_host = primaries[seg_id]
            m_host = (p_host + 1) % n_hosts
            mirrors.append(m_host)

        return primaries, mirrors

    @staticmethod
    def generate_unbalanced_spread(n_segments: int, n_hosts: int, 
                                  skew: float = 0.3) -> Tuple[List[int], List[int]]:
        """Generate UNBALANCED SPREAD configuration
        
        Unbalanced: Skewed primary distribution
        Spread: Mirrors distributed across all available hosts
        
        Args:
            skew: Fraction of segments to transfer (0.3 = 30% imbalance)
        """
        base = n_segments // n_hosts

        # Create skewed distribution
        loads = [base] * n_hosts

        # Transfer segments from last hosts to first hosts
        transfer = int(base * skew)
        if transfer > 0 and n_hosts >= 2:
            loads[0] += transfer
            loads[-1] -= transfer

        # Assign primaries
        primaries = []
        for host_id in range(n_hosts):
            primaries.extend([host_id] * loads[host_id])

        # Assign mirrors (spread pattern)
        mirrors = []
        for seg_id in range(n_segments):
            p_host = primaries[seg_id]
            
            # Available mirror hosts (not self)
            available = [h for h in range(n_hosts) if h != p_host]
            
            # Spread mirrors across available hosts
            m_host = available[seg_id % len(available)]
            mirrors.append(m_host)

        return primaries, mirrors

    @staticmethod
    def generate_multi_tier_unbalanced(n_segments: int, n_hosts: int, 
                                       heavy_hosts: int = 2,
                                       skew: float = 0.5) -> Tuple[List[int], List[int]]:
        """Generate multi-tier unbalanced configuration
        
        Creates tiers: some hosts heavily loaded, others lightly loaded
        
        Args:
            heavy_hosts: Number of hosts with heavy load
            skew: How much extra load heavy hosts get
        """
        assert heavy_hosts < n_hosts, "Need at least one light host"
        
        base = n_segments // n_hosts
        light_hosts = n_hosts - heavy_hosts
        
        # Calculate loads
        # Heavy hosts get more, light hosts get less
        extra_per_heavy = int((base * skew * light_hosts) / heavy_hosts)
        reduce_per_light = int((base * skew * light_hosts) / light_hosts)
        
        loads = []
        for host_id in range(n_hosts):
            if host_id < heavy_hosts:
                loads.append(base + extra_per_heavy)
            else:
                loads.append(max(1, base - reduce_per_light))
        
        # Adjust to match total segments exactly
        diff = n_segments - sum(loads)
        loads[0] += diff
        
        # Assign primaries
        primaries = []
        for host_id in range(n_hosts):
            primaries.extend([host_id] * loads[host_id])
        
        # Assign mirrors (spread pattern)
        mirrors = []
        for seg_id in range(n_segments):
            p_host = primaries[seg_id]
            available = [h for h in range(n_hosts) if h != p_host]
            m_host = available[seg_id % len(available)]
            mirrors.append(m_host)
        
        return primaries, mirrors

    @staticmethod
    def generate_random_unbalanced(n_segments: int, n_hosts: int,
                                  min_load_ratio: float = 0.5,
                                  spread: bool = True) -> Tuple[List[int], List[int]]:
        """Generate random unbalanced configuration
        
        Args:
            min_load_ratio: Minimum load as fraction of average (0.5 = at least 50% of average)
            spread: Use spread pattern for mirrors (otherwise grouped)
        """
        base = n_segments // n_hosts
        min_load = max(1, int(base * min_load_ratio))
        
        # Generate random loads
        loads = []
        remaining = n_segments
        
        for host_id in range(n_hosts - 1):
            # Random load between min and what's left
            max_load = remaining - min_load * (n_hosts - host_id - 1)
            load = random.randint(min_load, min(max_load, base * 2))
            loads.append(load)
            remaining -= load
        
        # Last host gets remainder
        loads.append(remaining)
        
        # Shuffle host assignment
        random.shuffle(loads)
        
        # Assign primaries
        primaries = []
        for host_id in range(n_hosts):
            primaries.extend([host_id] * loads[host_id])
        
        # Assign mirrors
        mirrors = []
        if spread:
            # Spread pattern
            for seg_id in range(n_segments):
                p_host = primaries[seg_id]
                available = [h for h in range(n_hosts) if h != p_host]
                m_host = available[seg_id % len(available)]
                mirrors.append(m_host)
        else:
            # Grouped pattern
            for seg_id in range(n_segments):
                p_host = primaries[seg_id]
                m_host = (p_host + 1) % n_hosts
                mirrors.append(m_host)
        
        return primaries, mirrors

    @staticmethod
    def generate_worst_case(n_segments: int, n_hosts: int) -> Tuple[List[int], List[int]]:
        """Generate worst-case unbalanced configuration
        
        All segments on first host, mirrors on second host
        """
        primaries = [0] * n_segments
        mirrors = [1 % n_hosts] * n_segments
        
        return primaries, mirrors

    @staticmethod
    def validate_config(primaries: List[int], mirrors: List[int], n_hosts: int) -> bool:
        """Validate that configuration is legal"""
        if len(primaries) != len(mirrors):
            
            return False
        
        for seg_id, (p, m) in enumerate(zip(primaries, mirrors)):
            if p < 0 or p >= n_hosts:
                
                return False
            if m < 0 or m >= n_hosts:
                
                return False
            if p == m:
                
                return False
        
        return True
```

</details>

```python
if __name__ == '__main__':
    conf = ConfigGenerator.generate_unbalanced_grouped(1000, 50)
    config = SolverConfig(1000, 40, 50, conf[0], conf[1],
                           strategy='grouped')
    solver = GreedySolver(config, run_improve=True, printing=True)

    solution, cost = solver.solve()
```

bimboterminator1 committed:
Initial

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,460 @@
+        elif self.options.target_segment_count > self.gparray.get_segment_count():
+            raise PlanningError("Expand is not supported yet")
+        elif self.options.target_segment_count == self.gparray.get_segment_count():
+            if self.options.skip_rebalance:
```
This check looks excessive, as we'll check again `self.options.skip_rebalance` right below and return the `plan`, which will be also `None` for the case `self.options.target_segment_count == self.gparray.get_segment_count()`

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,460 @@
+        plan.setTargetSegmentCount(self.options.target_segment_count)
+        return plan
+    
+    def removeSegPairFromArray(self, segPair: gparray.SegmentPair):
```
there is a mix of CamelCase and snake_case functions inside the same class. Let's keep the style inside a signle class consistent.

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,460 @@
+        
+        if not plan:
+            plan = Plan()
+        rebalance_moves = self.plan_rebalance()
```
From my point of view, here is a bit inconsistent code, because `plan_rebalance()` returns `List[LogicalMove]`, while `plan_shrink()` returns `ShrinkPlan`. I think that one could expect that all `plan_*` functions should return some derivative of `Plan` class.

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,460 @@
+
+        return moves
+    
+    def already_balanced(self, load):
```
Please add the return type

whitehawk commented:
For a configuration with  24 segments, 12 initial hosts and 6 target hosts:
```
if __name__ == '__main__':
    solver = BABSolver(
        n_segments=24,
        n_hosts_target=6,
        n_hosts_initial= 12, #host #3 is decommissioned
        initial_primary=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
        initial_mirror=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0],
        strategy='grouped',
        printing=True
    )

    solution, cost = solver.solve()
```
it took 300 seconds to compute:
```
======================================================================
SEARCH COMPLETE
======================================================================
Nodes explored:    110,119
Nodes pruned:      107,052
  Nogood prunes:   0
  Bound prunes:    33,413
  Load prunes:     1,946
Pruning rate:      97.2%
Time elapsed:      300.01s
Nodes/second:      367

Best cost:         26
======================================================================
```
which is too long I think. I'm afraid that for higher segment count (I think that hundreds could be valid number for a valid cluster) the compute time would be too long.

BTW, is there a way to check the exact output segment configuration besides this summary during the test?

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1063 @@
+                    if primary_load[under_host] >= self.L_target:
+                        continue  # This destination is now full
+                    
+                    dest = under_host
```
Shouldn't we take the first suitable one? Or why should we take the last one?

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1063 @@
+        self.L_target = solver.L_target
+        self.delta = 0
+        self.L_min = solver.L_target
+        self.L_max = solver.L_target
```
Add comments to the variables

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1063 @@
+        
+        # Convert to group structures (only if 2+ segments share primary)
+        mirror_groups = []
+        for primary_host in sorted(groups_by_primary.keys()):
```
In the test example from the description, `groups_by_primary.keys()` returns the key "default_factory" is this expected?

bimboterminator1 committed:
optimization

bimboterminator1 committed:
Change perspective

bimboterminator1 committed:
formatting

bimboterminator1 committed:
slight logic adjustment

whitehawk commented on gpMgmt/bin/gprebalance_modules/test/config.py:
```diff
@@ -0,0 +1,34 @@
+            self.encoding = None
+            return res
+        return wrapper
+    return inner
```
Please add new line

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,476 @@
+class ConfigurationEncoder:
+
+    @staticmethod
+    def encode_configutaion(gparray: gparray.GpArray, 
```
```suggestion
    def encode_configuration(gparray: gparray.GpArray, 
```

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,476 @@
+                            target_hosts: List[Host],
+                            strategy: str) -> Tuple[Encoding, HostMapping]:
+        """
+        The rebalance solvers work with abstract input segment configutation.
```
```suggestion
        The rebalance solvers work with abstract input segment configuration.
```
and please check other comments for typos.

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,476 @@
+        if not strat:
+            strat = self.get_mirroring_strat()
+        
+        if strat == 'spread' and expected_per_host > total_hosts - 1:
```
```suggestion
        if strat == SPREAD and expected_per_host > total_hosts - 1:
```

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,876 @@
+        
+        # Check 1: All segments assigned
+        if len(solution) != self.n_segments:
+            valid = False
```
I guess once we found invalid marker, there is no sence in further checks. So I think we can just `return False` at every place we assign  `valid = False`.

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,476 @@
+        # dry movements are planned here
+        conf, host_mapping = ConfigurationEncoder.encode_configutaion(self.virtual_gparray, self.target_hosts, strat)
+        id_to_host = {v: k for k, v in host_mapping.items()}
+        solution, cost = balancer.GreedySolver(*conf).solve()
```
`cost` is not used. Do we really need it? How are we going to use it?

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,876 @@
+    def improve(self, primary: List[int], mirror: List[int], 
+                iterations: int = 1000, time_limit: int = 60) -> Tuple[List[int], List[int]]:
+        """
+        Adaptive large neighborhood search with simulated annealing.
```
Before we dive deeply into the algorythm, a general question: why can't we use existing implementations for ALNS? It looks that at least one exists - https://github.com/N-Wouda/ALNS. Maybe more are presented...

bimboterminator1 replied:
1. My main motivation is reducing external dependencies, which may require specific versions of packages which could conflict with already present ones. 
2. Imho, existing framework will just bring better code organization, rather than better results. 
3. We could create a separate ticket for performing the comparative research of available methods, which will probably take long time.

whitehawk commented on gpMgmt/bin/gprebalance_modules/rebalance_commons.py:
```diff
@@ -1,121 +1,56 @@
+class HostStatus(IntEnum):
+    ACTIVE = 1
+    NEW = 2
+    DECOMMISSION = 3
```
If speaking about status of some host, better
```suggestion
    DECOMMISSIONED = 3
```

whitehawk commented on gpMgmt/bin/ggrebalance:
```diff
@@ -58,6 +58,20 @@ def parseargs():
     parser.add_option('-y', '--non-interactive-mode', dest="interactive", action='store_false', default=True,
                          help="non-interactive mode, do not require user input for confirmations, use defaults.")
+    parser.add_option('--target-hosts', dest='target_hosts', metavar='<target_hosts>',
+                      help='Set of hostnames where rebalance will distirbute segments to')
```
```suggestion
                      help='Set of hostnames where rebalance will distribute segments to')
```

whitehawk commented on gpMgmt/bin/ggrebalance:
```diff
@@ -58,6 +58,20 @@ def parseargs():
                       help='show this help message and exit.')
     parser.add_option('-y', '--non-interactive-mode', dest="interactive", action='store_false', default=True,
                          help="non-interactive mode, do not require user input for confirmations, use defaults.")
+    parser.add_option('--target-hosts', dest='target_hosts', metavar='<target_hosts>',
```
What about short aliaces here and below?

bimboterminator1 replied:
One-letter options could be counterintuitive in some aspects. How to encode --skip-rebalance? -sr? -s? -n?
or add-hosts: -a, -ah

whitehawk replied:
> One-letter options could be counterintuitive in some aspects. 

It is a matter of requirements. In current requirements there is a short alias `-H` is specifired for `--target-hosts` (refer to `REQ.1.1.2`). Some other params also have short aliases. So either we need to implement the requirement, or we need to change the requirement if it is proved to be wrong.

bimboterminator1 replied:
> It is a matter of requirement

Adjusted. Strict correspondance to requirements is not achieved for option (i renamed --no-rebalance to --skip-rebalance), i 'll handle it with architecht.

bimboterminator1 committed:
fix dict

bimboterminator1 committed:
.pop()

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,876 @@
+            if self.target_primary_load > self.n_hosts_target - 1:
+                raise ValueError("Cannot follow spread mirroring strategy")
+    
+    def solve(self) -> Tuple[Dict[int, Tuple[int, int]], int]:
```
Can we replace `Tuple[Dict[int, Tuple[int, int]], int]` with some object? For ex. add class `Solution` or smth similar. Reading code with such nested  Tuple->Dict->Tuple constructions is complicated.

bimboterminator1 replied:
I'll add type aliases, adding extra objects wrapping base types seems unneccesary to me. Critique is welcome

whitehawk replied:
Another example :
```
Encoding = Tuple[int, int, int, List[int], List[int], str]
```
this (and all similar cases) are hard to read and are error prone. It is not self-documenting - I have to crawl over other places of code to guess which value is responcible for. Maintainability and encapsulation suffer, such approach violates code clarity principles.

whitehawk commented on gpMgmt/bin/gprebalance_modules/test/config.py:
```diff
@@ -0,0 +1,34 @@
+from gppylib.gparray import Segment, GpArray
+from ..planner import ConfigurationEncoder, Planner
+
+def initGparrayFromFile(basename):
```
why can't we use existing `initFromFile()` instead of this new function?

bimboterminator1 replied:
we can't since for some configurations it could throw errors.

whitehawk replied:
Would it is reasonable to modify `initFromFile()` to handle such configurations instead of adding a separate function?

bimboterminator1 replied:
I remembered the issue: original function can't handle  # comments in the configuration file. I don't know how necessaary it is to support it, but I see nothing criminal in adding little separate function used in tests only.

whitehawk commented on gpMgmt/bin/gprebalance_modules/rebalance_commons.py:
```diff
@@ -1,121 +1,61 @@
+    def __str__(self):
+        pass
+
+class SegmentSize:
```
It looks not used. Do we need it right now?

bimboterminator1 replied:
Why not to just declare it as a concept/template for future work?

whitehawk replied:
If you are confident that it will be required in future, than it may be Ok. Please add a comment with some disclaimer in this case for the code.

bimboterminator1 committed:
fix

bimboterminator1 commented:
> It looks that behave tests are still broken

Now some tests pass, hovewer in some cases they will fail in the following way:

The ggrebalance fails at one of the steps after we succesfully shrink the segments (on_enter_STATE_SHRINK_CATALOG_STARTED and further) and gets to the STATE_ERROR. The logic in Planner.plan() can't definetely determine that we still running shrink (since options.target_segment_count == gparray.get_segment_count()). The proper handling of this stuff should be done via general state machine, not by the planning code. And preserving current behave tests is redundant.

whitehawk commented:
> > It looks that behave tests are still broken
> 
> Now some tests pass, hovewer in some cases they will fail in the following way:
> 
> The ggrebalance fails at one of the steps after we succesfully shrink the segments (on_enter_STATE_SHRINK_CATALOG_STARTED and further) and gets to the STATE_ERROR. The logic in Planner.plan() can't definetely determine that we still running shrink (since options.target_segment_count == gparray.get_segment_count()). The proper handling of this stuff should be done via general state machine, not by the planning code. And preserving current behave tests is redundant.

I do not like the idea that the PR will leave an existing test suite in a broken state.
I would prefer to have some simple temp code that will help to handle this case (for ex,. check for `rebalance_schema` existence, and if there is schema presented - do not go into planning or smth similar) together with TODOs marks, than leaving the test suite broken until some future PR (that is not mandatory to be the very next) and exposing us to risks of breaking existing functionality as we can't run the regression.

bimboterminator1 committed:
fix

bimboterminator1 committed:
fix2

bimboterminator1 commented:
> I would prefer to have some simple temp code that will help to handle this case

changed the code, tests should pass

> The issue looks still relevant:

changed

bimboterminator1 committed:
warn about planning time

bimboterminator1 committed:
short options

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,502 @@
+                    lines.append(f"        Port:     {seg_pair.primaryDB.port}")
+                
+                # Mirror segment
+                if seg_pair.primaryDB:
```
`primaryDB` - > `mirrorDB`?

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,502 @@
+        else:
+            lines.append("  No segments to remove.")
+        
+        # Moves section
```
It looks to me that this block has much logic intersection with the superclass `__str__`. I guess that any child class would have section describing moves, so I think it is better to move it to the super class in order to reduce code duplication.

bimboterminator1 committed:
add typing to unit test

whitehawk commented on gpMgmt/bin/gprebalance_modules/test/test_unit_solver.py:
```diff
@@ -0,0 +1,266 @@
+    def test_validity_small_grouped_unbalanced(self):
+        conf = self.encoding[0]
+
+        solver = solver = GreedySolver(conf, run_improve=False)
```
Is there any reason for double assignement `solver = solver = ...`? Or typo?

bimboterminator1 committed:
remove timeout exception

bimboterminator1 committed:
change unit test cost

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1525 @@
+    n_segments: int
+    n_hosts_target: int
+    n_hosts_initial: int
+    initial_primariy: List[HostId]
```
`initial_primariy` -> `initial_primary`?

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1525 @@
+        # Current primary load for each target host.
+        current_load = [0] * self.n_hosts_target
+        
+        for must_move, _, seg_id in segment_order:
```
`must_move` is not used in the loop

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1525 @@
+        else:  # 'any'
+            self._assign_mirrors_any(primary, mirror, mirror_load)
+        
+        return mirror
```
Please make consistent new-line padding between functions.

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1525 @@
+            )
+            segment_order.append((must_move, -initial_load[orig_host], i))
+        
+        segment_order.sort()
```
> must_move first

I get a list like this after this step.
`[(False, -2, 0), (False, -2, 1), (False, -2, 2), (False, -2, 3), (True, 0, 4), (True, 0, 5)]`

bimboterminator1 replied:
Yep, missed that. It does not affect the solution validity (no segments must be at hostid >= n_hosts_target),  only the initial solution and cost.

whitehawk replied:
I thought that was the intention - to process firstly the segments that *do not* require movements, filling them into the result `primary` list at their original places if possible. And only then add to the `primary` segments that require movements or their original host is overloaded.

But in this case I think we do not need to negate the `initial_load`:
```
            segment_order.append((must_move, initial_load[orig_host], i))

        segment_order.sort()
```

what do you think?

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1525 @@
+            3. Swap another already-assigned group with current in
+            case of conflict when the only mirror host is primary one.
+        """    
+        # Count how many segments in this group "prefer" each original mirror.
```
is mirror a segment or a host in this context?

bimboterminator1 replied:
Host is assumed in most similar places

bimboterminator1 committed:
formatting

bimboterminator1 committed:
sort primaries

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+            
+            # Try to keep on original host if possible
+            if (orig_host < self.n_hosts_target and 
+                current_load[orig_host] < self.target_primary_load):
```
This condition looks very similar to the one above (but with negation):
```
            must_move = (
                orig_host >= self.n_hosts_target
                or initial_load[orig_host] > self.target_primary_load
            )
```
Maybe we can move the check into a separate function, smth like 
```
def move_is_required(self, host, load) -> bool:
    return host >= self.n_hosts_target or load[host] > self.target_primary_load
```
and use it 
```
...
segment_order.append((move_is_required(orig_host, initial_load), -initial_load[orig_host], i))
...
if move_is_required(orig_host, current_load):
                host = orig_host
            else:
                # Find least loaded host
                host = min(range(self.n_hosts_target), key=lambda h: current_load[h])
```

bimboterminator1 replied:
Nice proposal,  but 2 conditions are not truly opposite. Added one simple function for is_decomissioned

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+        primary_to_mirror: Dict[HostId, HostId] = {}
+
+        # Process larger groups first for better balance
+        sorted_groups = sorted(groups.items(), key=lambda item: len(item[1]), reverse=True)
```
`groups` are made from `primary` list. My understanding is that by this moment, we've done initial placement of primaries, so every host contains equal number of segments (equal to `target_primary_load`). Therefore, all groups should have equal number of elements. So what is the purpose of this sorting?

bimboterminator1 replied:
Removed

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+            best_mirror = self._select_group_mirror(
+                p_host=p_host,
+                segs=segments,
+                group_size=group_size,
```
`group_size` is calculated from `segments`, and we pass both  `group_size` and `segments` as parameters. Why can't we get the `group_size` inside `_select_group_mirror()` and eliminate one of the parameters?

bimboterminator1 replied:
eliminated

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+        return mirror
+
+    def _assign_mirrors_grouped(self,
+                                primary: List[HostId],
```
`primary` parameter is not used in the function

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+            )
+            # Assign the group to chosen mirror.
+            for seg in segments:
+                mirror[seg] = best_mirror
```
When we are talking about mirror, usually we mean one particular mirror **segment**. But, for ex, here `mirror` is a **list of hosts**. `best_mirror` is a **host**. All different entities. I think readability should be improved here, and _in other similar places and functions_. Names should better reflect the meaning of variables. For ex. smth like:


```suggestion
                mirror_seg_to_host_mapping[seg] = best_mirror_host
```

bimboterminator1 replied:
Changed  some moments for understanding, refactoring the whole code multiple times is time consuming activity, which is often useful, but may bring sometimes more hidden errors. Mostly, I disagree with such formalism specifically for current situation

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+        """
+        Pick a mirror host for a primary group (grouped strategy),
+        with the following priority:
+            1. Most-voted original mirror host (with capacity).
```
'voted' is not a widely used term for a host. Please add a description what do you mean by this term at its first appearence.

bimboterminator1 replied:
Changes for 'used'. The meaning is unterstandable from context.

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+                groups=groups,
+                mirror=mirror)
+            if best_mirror is None:
+                raise RuntimeError("Unable tp assign mirror host in grouped strategy")
```
We have checks for the initial conditions:
```
        if self.n_hosts_target < 2:
            raise ValueError("Cannot balance to single host")
        
        if self.n_segments % self.n_hosts_target != 0:
            raise ValueError(f"Cannot evenly distribute {self.n_segments}"
                             f"segments across {self.n_hosts_target} hosts")
```
So, by this point it is guaranteed that all target hosts will have the same number of segments, and number of target hosts > 1. Correct me if I'm wrong, but for such pre-conditions it is **always possible** to get at least one valid grouped-mirror configuration - I suppose, a straight-forward shift of segment should always work:
```
primaries = [0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3]
mirrors = [1, 1, 1, 2, 2, 2, 3, 3, 3, 0, 0, 0]
```

In this case, if we are embedding the possibility to get a runtime failure of not finding a valid configuration, that reveals that there is something wrong with our approach.

bimboterminator1 replied:
Yep, it's always possible. But the approach won't be changed until it will be proved cardinally wrong and contradictory, sorry. Current variant can give better options than straighforward shift/round robin in some cases. 
For this configuration (added new host 3):
```
 config = SolverConfig(12, 4, 3, [0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2], [2, 2, 2, 2, 0, 0, 0, 0, 1, 1, 1, 1],
                           strategy='grouped')
  solver = GreedySolver(config, run_improve=False, printing=True)

primary [0, 0, 0, 3, 1, 1, 1, 3, 2, 2, 2, 3]
mirror [2, 2, 2, 0, 3, 3, 3, 0, 1, 1, 1, 0]
mirror moves = 5

straighforward mirror [1, 1, 1, 0, 2, 2, 2, 0, 3, 3, 3, 0]
mirror moves = 11
```
simple 0-1 1-2 2-3 3-0 assignment leads to more mirror moves than approach when we want keeping the original placement. In some cases straightforward assignment could be better, nobody argues, current variant keeps balance between code complexity and intelligence during assignment. Other things being equal, it is expected to recieve nearly the same cost as for straightforward assignment, maybe better. In other words, following the **heuristic of prioritizing original mirror host** comes in handy here. Now for grouped cases the optimal solution sometimes is even found.

Code is readable enough to understand. RuntimeError is added mostly for style, rather than for covering pitfalls, which does not exist. Also they could prevent some development/debugging errors in future.

whitehawk replied:
> But the approach won't be changed until it will be proved cardinally wrong and contradictory

I'm not saying we need for 100% to change the approach. I'm saying that having an error in production code, when we know for sure there is an existing solution, is a bit awkward - almost as if a `float sin(float x)` would throw an error for particular `x` values, if it can't calculate `sin` fast enough.
What about, instead of the error, add a warning and do fallback to dumb, non-optimal, but robust straighforward shift/round robin? That will reduce the possibility that the end user will face malfunction of the tool, if he has some exotic cluster configuration that we missed during testing.

bimboterminator1 replied:
>What about, instead of the error, add a warning and do fallback to dumb, non-optimal, but robust straighforward shift/round robin?

Sounds fine, let's do it

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+        elif self.strategy == 'spread':
+            self._assign_mirrors_spread(primary, mirror, mirror_load, groups)
+        else:  # 'any'
+            self._assign_mirrors_any(primary, mirror, mirror_load)
```
According to requirements:
```
REQ.6.3.3	ggrebalance should support group and spread mirroring strategies.
```
so I do not see reason in adding additional 'mirrors_any' strategy. It looks to be additional code that we need to support and test for no sence.

bimboterminator1 replied:
Ok, i'll block it from ggrebalance side and remove it from solver

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/rebalance_schema.py:
```diff
@@ -0,0 +1,120 @@
+    def rebalanceSchema(self, target_segment_count: int) -> None:
+        # Before rebalancing check if the tables are already rebalanced
+        # (in case we re-enter after interruption that happened after COMMIT but before new state)
+        if get_table_distr_segment_count(self.conn, self.schema_name, self.rebalance_status) > target_segment_count:
```
What if the table is not located on all segments? Should this be handled in this patch?

bimboterminator1 replied:
> Should this be handled in this patch?

Not really. This code has existed earlier. 

> What if the table is not located on all segments?

According to current logic it will be reorganized to target_segment_count.

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1531 @@
+                            best_cost=initial_cost)
+        
+        if self.solver.printing:
+            print(f"Initial cost: {state.best_cost}")
```
Should we use gplog?

bimboterminator1 replied:
The gppylib logger passing into this code violates its isolation from client code semantics. The print calls are used for debug purposes only, such info is unnecessary in main log, i assume.

bimboterminator1 committed:
improvements

bimboterminator1 committed:
improvements

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1487 @@
+
+    Hosts:
+        initial hosts: 0 .. n_hosts_initial-1
+        target hosts:  0 .. n_hosts_target-1 (subset of initial hosts)
```
Are target hosts always a subset of initial hosts? If I try to expand from 1 host to 2, it also seems to create a valid config:
```
    config = SolverConfig(10,
                          2,
                          1,
                          [0,0,0,0,0,0,0,0,0,0],
                          [0,0,0,0,0,0,0,0,0,0],
                          strategy='grouped')
    solver = GreedySolver(config, run_improve=True, printing=True)
    solution, cost = solver.solve()
    prims: Dict[HostId, List[ContentId]] = defaultdict(list)
    mirrs: Dict[HostId, List[ContentId]] = defaultdict(list)
    for seg, hosts in solution.items():
        prims[hosts[0]].append(seg)
        mirrs[hosts[1]].append(seg)
    prims = dict(sorted(prims.items()))
    mirrs = dict(sorted(mirrs.items()))
    print(prims)
    print(mirrs)
```
```
Initial cost: 10
Final cost: 10
{0: [0, 1, 2, 3, 4], 1: [5, 6, 7, 8, 9]}
{0: [5, 6, 7, 8, 9], 1: [0, 1, 2, 3, 4]}

```

bimboterminator1 replied:
The correct configuration for expansion is target_hosts = 2 and initial_hosts 2 as well (in segment mapping we still may have all zeros) . Everything should follow the encoding rules described in ConfigurationEncoder.encode_configuration

bimboterminator1 committed:
revert

bimboterminator1 committed:
add fallback to straightforward

whitehawk commented on gpMgmt/bin/gprebalance_modules/planner.py:
```diff
@@ -1,19 +1,472 @@
+    def __str__(self):
+        lines = []
+
+        lines.append(f"\n{'BALANCE MOVES':-^80}")
```
Nice to have (not for current phase, but for later patch) - besides moves, to show also the final cluster configuration.

bimboterminator1 replied:
Agree

whitehawk commented on gpMgmt/bin/gprebalance_modules/solver.py:
```diff
@@ -0,0 +1,1526 @@
+            )
+            # Fallback to straightforward assignment (shouldn't happen)
+            if best_mirror_host is None:
+                self._fill_naive_grouped(primary_mapping, mirror_mapping)
```
Maybe add here a warning about fallback - smth like 'fallback happened, please contact developers and provide your initial and target cluster configuration for the investigation to make the product better'.

bimboterminator1 replied:
1. I venture to say that it will not fallback.
2. The gppylib logger passing into this code violates its isolation from client code semantics. And doing that via print() is meaningless.

whitehawk commented on gpMgmt/bin/gprebalance_modules/test/data/config_generator.py:
```diff
@@ -0,0 +1,616 @@
+        print("\n  For help:")
+        print("    python generator.py -h")
+    else:
+        main()
```
please add new line at the file end

whitehawk commented on gpMgmt/bin/gprebalance_modules/test/test_unit_solver.py:
```diff
@@ -0,0 +1,689 @@
+    
+    @getEncoding('1000_50_unbalanced_grouped', 'grouped', None, None, None)
+    def test_validity_large_grouped_unbalanced(self):
+        conf = self.encoding[0]
```
It looks that many test here follow the same structure:
```
        conf = self.encoding[0]

        solver = GreedySolver(conf, run_improve=<...>)

        solution, cost = solver.solve()

        self.assertEqual(cost, <...>)

        self._validate_solition(solution, solver)
```
and the only difference is `run_improve` and expected cost value. What about moving the duplciated code into a function, smth like `perform_test(run_improve, expected_cost)`, so the test body will be reduced, for example, to 
```
    perform_test(true, 140)
```

bimboterminator1 replied:
Done

whitehawk commented on gpMgmt/bin/gprebalance_modules/test/test_unit_solver.py:
```diff
@@ -0,0 +1,689 @@
+                    return False
+        return True
+    
+    def _validate_solition(self, solution: Solution, 
```
Before we get here, will not the assert 
```
assert(self._validate_solution(solution))
```
inside `solver.solve()` happen, if there is some issue with the solution? In that's true, this code is not reachable.

bimboterminator1 replied:
Even the validation code is not reached, the test fill fail with ERROR showing the exception reason. It's ok for me.

whitehawk replied:
So why should we keep the unreachable code?

bimboterminator1 replied:
It's executed for positive scenarios, it's ok to have it to describe the behaviour. Moreover, what if the solver's own validation code will be broken and assert won't fire. I'll keep it

bimboterminator1 committed:
add new line

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/test/test_unit_solver.py:
```diff
@@ -0,0 +1,502 @@
+    def perform_run(self, run_improve:bool, cost:int):
+        conf = self.encoding[0]
+        solver = GreedySolver(conf, run_improve=run_improve)
+        solution, cost = solver.solve()
```
`cost` variable is overwrited.

bimboterminator1 replied:
Yep, fixed

KnightMurloc commented on gpMgmt/bin/gprebalance_modules/test/test_unit_solver.py:
```diff
@@ -0,0 +1,502 @@
+        conf = self.encoding[0]
+        solver = GreedySolver(conf, run_improve=run_improve)
+        solution, cost = solver.solve()
+        self.assertEqual(cost, cost)
```
`cost` variable is overwrited.

bimboterminator1 committed:
fix overwritting

KnightMurloc commented:
`gprebalance balances cluster after expansion on fixed hosts` test is failing. Is it expected?
